lib/factor-rules.js

```diff
@@ -0,0 +1,373 @@
+import {isPolynomialTerm, getCoefficient, getVariableFactors, getCoefficientsAndConstants} from './rules/collect-like-terms'
+import {clone, getRanges, flattenOperands} from './utils'
+
+const defineRuleString = (matchPattern, rewritePattern, constraints) => {
+    const matchAST = parse(matchPattern)
+    const rewriteAST = parse(rewritePattern)
+
+    traverse(matchAST, {
+        leave(node) {
+            delete node.loc
+        }
+    })
+
+    traverse(rewriteAST, {
+        leave(node) {
+            delete node.loc
+        }
+    })
+
+    return definePatternRule(matchAST, rewriteAST, constraints)
+}
+
+// FACTOR
+
+// TODO: add minus case
+// assume monic polynomial
+// TODO: add min and max to evaluate
+
+// e.g. x^2 + x^5 + x^6 -> x^2(x^0 + x^3 + x^4)
+export const FACTOR_SYMBOL = defineRuleString('#a^#b_0 + ...', '#a^#eval(min(#b_0, ...)) (#a^#eval(#b_0 - min(#b_0, ...)) + ...)', {b: query.isNumber})
+
+// TODO: add restrictions (#c_0 divisible by 2 && #a_0 is perfect square)
+
+// e.g. 4x^2 - 9y^2 -> (2x)^2 - (3y)^2
+export const FACTOR_DIFFERENCE_OF_SQUARES_HELPER =
+  defineRuleString('#a #b^#c - #d #e^#f', '(#eval(nthRoot(#a)) #b^(#eval(#c/2)))^2 - (#eval(nthRoot(#d)) #e^(#eval(#f/2)))^2')
+
+// e.g. (2x)^2 - (3y)^2 -> (2x + 3y)(2x - 3y)
+export const FACTOR_DIFFERENCE_OF_SQUARES =
+  defineRuleString('#a^2 - #b^2', '(#a + #b)(#a - #b)')
+
+
+export const hello = (num, root = 2, precision = 12) => {
+  // e.g 2^-3 = 1/(2^3)
+  const inv = root < 0
+  if (inv) {
+    root = -root
+  }
+
+  if (root === 0) {
+    throw new Error('Root must be non-zero')
+  }
+  if (num < 0 && (Math.abs(root) % 2 !== 1)) {
+    throw new Error('Root must be odd when a is negative.')
+  }
+
+  // Edge cases zero and infinity.
+  // e.g 0^3 = 0, 0^-3 = Infinity
+  if (num === 0) {
+    return inv ? Infinity : 0
+  }
+
+  if (num === 1) {
+    return 1
+  }
+
+  if (!isFinite(num)) {
+    return inv ? 0 : num
+  }
+
+  // Source: https://rosettacode.org/wiki/Nth_root#JavaScript
+  const n = root
+  const prec = precision
+
+  let x = 1 // Initial guess
+  for (let i = 0 ; i < prec ; i++) {
+    x = 1 / n * ((n - 1) * x + (num / Math.pow(x, n - 1)))
+  }
+
+  return inv ? 1 / x : x
+}
+
+// TODO: handle fractional coefficients
+// TODO: isPolynomialTerm should handle fractional coeffs
+
+// e.g. 4x^2 + 12x + 9 -> (2x + 3)^2
+export const FACTOR_PERFECT_SQUARE_TRINOMIALS = defineRule(
+  (node) => {
+    const isFactorable = canApplyRule(FACTOR_SUM_PRODUCT_RULE, node)
+    if (isFactorable) {
+      const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+      return print(result.args[0]) === print(result.args[1])
+        ? {node} : null
+    }
+  },
+
+  (node) => {
+    const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+    return build.pow(result.args[0], build.number(2))
+  }
+)
+
+const getExponent = (node) => {
+    if (query.isNumber(node)) {
+        return 0
```

Comment:
Oh, this `getExponent` function is to get the degree of x in a polynomial. So for instance, `6x^2 + 3x + 1`, `getExponent` on 1 would return 0 because it is `1x^0`.